lib/user.py

```diff
@@ -0,0 +1,178 @@
+import json
+import hashlib
+import uuid
+
+from lib.secure_variable import SecureVariable
+from lib.db import DB
+
+class User():
+
+    def __init__(self, user_id: int):
+        user = DB().fetch_one("""
+                SELECT id, name, capabilities
+                FROM users
+                WHERE id = %s
+                """, params=(user_id, ))
+        if not user:
+            raise RuntimeError(f"User with id {user_id} not found in database")
+
+        self._id = user[0]
+        self._name = user[1]
+        self._capabilities = user[2]
+
+
+    def __repr__(self):
+        values = self.__dict__.copy()
+        del values['_id']
+        return str(values)
+
+    def update(self):
+        DB().query("""
+            UPDATE users
+            SET capabilities = %s
+            WHERE id = %s
+            """, params=(json.dumps(self._capabilities), self._id, ))
+
+    def can_use_machine(self, machine_id: int):
+        return machine_id in self._capabilities['machines']
+
+    def can_use_route(self, route: str):
+        return route in self._capabilities['api']['routes']
+
+    def can_schedule_job(self, schedule_mode: str):
+        return schedule_mode in self._capabilities['jobs']['schedule_modes']
+
+
+    def has_api_quota(self, route: str):
+        if route in self._capabilities['api']['quotas']:
+            return self._capabilities['api']['quotas'][route] > 0
+        return True # None means infinite amounts
+
+    def deduct_api_quota(self, route: str, amount: int):
+        if route in self._capabilities['api']['quotas']:
+            self._capabilities['api']['quotas'][route] -= amount
+            self.update()
+
+    def has_measurement_quota(self, machine_id: int):
+        machine_id = str(machine_id) # json does not support integer keys
+        if machine_id in self._capabilities['measurement']['quotas']:
+            return self._capabilities['measurement']['quotas'][machine_id] > 0
+        return True # None means infinite amounts
+
+    def deduct_measurement_quota(self, machine_id: int, amount: int):
+        machine_id = str(machine_id)  # json does not support integer keys
+        if machine_id in self._capabilities['measurement']['quotas']:
+            self._capabilities['measurement']['quotas'][machine_id] -= amount
+            self.update()
+
+    @classmethod
+    def authenticate(cls, token: SecureVariable | None, silent=False):
+        sha256_hash = hashlib.sha256()
+        if token is None or token.get_value() is None:
+            sha256_hash.update("DEFAULT".encode('UTF-8'))
+        else:
+            sha256_hash.update(token.get_value().encode('UTF-8'))
+
+        user = DB().fetch_one("""
+                SELECT id, name
+                FROM users
+                WHERE token = %s
+                """, params=((sha256_hash.hexdigest()), ))
```

Comment:
We need to make sure that every SQL connection is encrypted otherwise the tokens are exposed

Reply:
Let's have a separate discussion how far we want to go here. Technically GMT itself has no SSL encryption to begin with. It is only there because we have Cloudflare in front of it.
Also we have no in-memory encryption which is also something that could be helpful ...

Moved to a discussion: https://github.com/green-coding-solutions/green-metrics-tool/discussions/918

Please follow up there